The job **`x86_64-gnu-llvm-16`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7472176266/job/20333828283) [(plain)](https://github.com/rust-lang/rust/commit/281ceb2bd2b2dd5f6ac201f585a2de2d21a41250/checks/20333828283/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Built container sha256:9c3c93a371e5aed5c18185b24f130d95d5140dbd72a9b325e7b6b49e521a4faa
Looks like docker image is the same as before, not uploading
https://ci-caches.rust-lang.org/docker/7ebc15c01a233894034d277c8cce4e949f4e7791f66b4727c8fb6e058a0b8171d6152e1441d677cef0653843ceeee469c097b8699b2bb74249e674f6aa1a8813
sha256:9c3c93a371e5aed5c18185b24f130d95d5140dbd72a9b325e7b6b49e521a4faa
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-16]
##[group]Clock drift check
  local time: Wed Jan 10 08:29:23 UTC 2024
  network time: Wed, 10 Jan 2024 08:29:23 GMT
  network time: Wed, 10 Jan 2024 08:29:23 GMT
##[endgroup]
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-16', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'build.optimized-compiler-builtins', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-16/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
##[endgroup]
Testing GCC stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished release [optimized] target(s) in 1.36s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --no-default-features --mini-tests --std-tests`
Using system GCC
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
F

failures:

---- compiler/rustc_index/src/vec.rs - vec::IndexVec (line 22) stdout ----
error[E0432]: unresolved imports `rustc_index::IndexVec`, `rustc_index::newtype_index`
  |
  |
6 | use rustc_index::{IndexVec, newtype_index};
  |                   ^^^^^^^^  ^^^^^^^^^^^^^ no `newtype_index` in the root
  |                   no `IndexVec` in the root
  |                   no `IndexVec` in the root
error: doctest failed, to rerun pass `-p rustc_index --doc`
  = help: consider importing this struct instead:
          ::rustc_index::IndexVec

error: cannot determine resolution for the macro `newtype_index`
error: cannot determine resolution for the macro `newtype_index`
##[error] --> compiler/rustc_index/src/vec.rs:27:1
  |
8 | newtype_index! {
  | ^^^^^^^^^^^^^
  |
  = note: import resolution is stuck, try simplifying macro imports

error[E0412]: cannot find type `MyIdx` in this scope
   |
   |
12 | let my_index_vec: IndexVec<MyIdx, u32> = IndexVec::from_raw(vec![0,1,2,3]);
   |
help: you might be missing a type parameter
   |
   |
4  | fn main() { #[allow(non_snake_case)] fn _doctest_main_compiler_rustc_index_src_vec_rs_22_0<MyIdx>() {


error[E0412]: cannot find type `MyIdx` in this scope
   |
   |
13 | let idx: MyIdx = MyIdx::from_u32(2);

error[E0433]: failed to resolve: use of undeclared type `MyIdx`
##[error]  --> compiler/rustc_index/src/vec.rs:32:18
   |
   |
13 | let idx: MyIdx = MyIdx::from_u32(2);
   |                  ^^^^^ use of undeclared type `MyIdx`
error: aborting due to 5 previous errors

Some errors have detailed explanations: E0412, E0432, E0433.
For more information about an error, try `rustc --explain E0412`.
```

</details>

--- a/compiler/rustc_index/src/vec.rs
+++ b/compiler/rustc_index/src/vec.rs
@@ -17,27 +17,15 @@ use crate::{Idx, IndexSlice};
 /// While it's possible to use `u32` or `usize` directly for `I`,
 /// you almost certainly want to use a [`newtype_index!`]-generated type instead.
 ///
-/// This allows to index the IndexVec with the new index type:
+/// This allows to index the IndexVec with the new index type
 ///
-/// ```
-/// use crate as rustc_index;
-/// use rustc_index::{IndexVec, newtype_index};
-///
-/// newtype_index! {
-///   pub struct MyIdx {}
-/// }
-///
-/// let my_index_vec: IndexVec<MyIdx, u32> = IndexVec::from_raw(vec![0,1,2,3]);
-/// let idx: MyIdx = MyIdx::from_u32(2);
-/// assert_eq!(my_index_vec[idx], 2);
-/// ```
 ///
 /// [`newtype_index!`]: ../macro.newtype_index.html
 #[derive(Clone, PartialEq, Eq, Hash)]
 #[repr(transparent)]
 pub struct IndexVec<I: Idx, T> {
     pub raw: Vec<T>,
-    _marker: PhantomData<fn(&I)>,
+    _marker: PhantomData<I>,
 }
 
 impl<I: Idx, T> IndexVec<I, T> {
@@ -77,7 +65,7 @@ impl<I: Idx, T> IndexVec<I, T> {
         IndexVec::from_raw(vec![elem; universe.len()])
     }
 
-    /// Creates a new `IndexVec` with n copies of `elem`
+    /// Creates a new IndexVec
     #[inline]
     pub fn from_elem_n(elem: T, n: usize) -> Self
     where

--- a/compiler/rustc_index/src/vec.rs
+++ b/compiler/rustc_index/src/vec.rs
@@ -12,9 +12,25 @@ use std::vec;
 use crate::{Idx, IndexSlice};
 
 /// An owned contiguous collection of `T`s, indexed by `I` rather than by `usize`.
+/// Its purpose is to avoid mixing indexes.
 ///
 /// While it's possible to use `u32` or `usize` directly for `I`,
 /// you almost certainly want to use a [`newtype_index!`]-generated type instead.
+///
+/// This allows to index the IndexVec with the new index type:
+///
+/// ```
+/// use crate as rustc_index;
+/// use rustc_index::{IndexVec, newtype_index};
+///
+/// newtype_index! {
+///   pub struct MyIdx {}
+/// }
+///
+/// let my_index_vec: IndexVec<MyIdx, u32> = IndexVec::from_raw(vec![0,1,2,3]);
+/// let idx: MyIdx = MyIdx::from_u32(2);
+/// assert_eq!(my_index_vec[idx], 2);
+/// ```
 ///
 /// [`newtype_index!`]: ../macro.newtype_index.html
 #[derive(Clone, PartialEq, Eq, Hash)]
@@ -25,11 +41,13 @@ pub struct IndexVec<I: Idx, T> {
 }
 
 impl<I: Idx, T> IndexVec<I, T> {
+    /// Constructs a new, empty `IndexVec<I, T>`.
     #[inline]
     pub const fn new() -> Self {
         IndexVec::from_raw(Vec::new())
     }
 
+    /// Constructs a new `IndexVec<I, T>` from a `Vec<T>`
     #[inline]
     pub const fn from_raw(raw: Vec<T>) -> Self {
         IndexVec { raw, _marker: PhantomData }
@@ -59,6 +77,7 @@ impl<I: Idx, T> IndexVec<I, T> {
         IndexVec::from_raw(vec![elem; universe.len()])
     }
 
+    /// Creates a new `IndexVec` with n copies of `elem`
     #[inline]
     pub fn from_elem_n(elem: T, n: usize) -> Self
     where

--- a/compiler/rustc_index/src/vec.rs
+++ b/compiler/rustc_index/src/vec.rs
@@ -17,15 +17,13 @@ use crate::{Idx, IndexSlice};
 /// While it's possible to use `u32` or `usize` directly for `I`,
 /// you almost certainly want to use a [`newtype_index!`]-generated type instead.
 ///
-/// This allows to index the IndexVec with the new index type
-///
-///
+/// This allows to index the IndexVec with the new index type.
 /// [`newtype_index!`]: ../macro.newtype_index.html
 #[derive(Clone, PartialEq, Eq, Hash)]
 #[repr(transparent)]
 pub struct IndexVec<I: Idx, T> {
     pub raw: Vec<T>,
-    _marker: PhantomData<I>,
+    _marker: PhantomData<fn(&I)>,
 }
 
 impl<I: Idx, T> IndexVec<I, T> {
@@ -35,7 +33,7 @@ impl<I: Idx, T> IndexVec<I, T> {
         IndexVec::from_raw(Vec::new())
     }
 
-    /// Constructs a new `IndexVec<I, T>` from a `Vec<T>`
+    /// Constructs a new `IndexVec<I, T>` from a `Vec<T>`.
     #[inline]
     pub const fn from_raw(raw: Vec<T>) -> Self {
         IndexVec { raw, _marker: PhantomData }
@@ -65,7 +63,7 @@ impl<I: Idx, T> IndexVec<I, T> {
         IndexVec::from_raw(vec![elem; universe.len()])
     }
 
-    /// Creates a new IndexVec
+    /// Creates a new IndexVec with n copies of the `elem`.
     #[inline]
     pub fn from_elem_n(elem: T, n: usize) -> Self
     where
@@ -92,6 +90,7 @@ impl<I: Idx, T> IndexVec<I, T> {
         IndexSlice::from_raw_mut(&mut self.raw)
     }
 
+    /// Pushes an element to the array returning the index where it was pushed to.
     #[inline]
     pub fn push(&mut self, d: T) -> I {
         let idx = self.next_index();

